documentation/proposals/Proposal - Axis Input Devices.md

```diff
@@ -0,0 +1,624 @@
+        new(21, AxisTrait.Position, "Touchpad X1"),
+        new(22, AxisTrait.Position, "Touchpad Y1"),
+        new(23, AxisTrait.Position, "Touchpad X2"),
+        new(24, AxisTrait.Position, "Touchpad Y2"),
+        new(25, AxisTrait.Position, "Touchpad X3"),
+        new(26, AxisTrait.Position, "Touchpad Y3"),
+        new(27, AxisTrait.Position, "Touchpad X4"),
+        new(28, AxisTrait.Position, "Touchpad Y4"),
+
+        // select/start/home
+        new(29, AxisTrait.Binary | AxisTrait.LeftSide, "Share"),
+        new(30, AxisTrait.Binary | AxisTrait.RightSide, "Start"),
+        new(31, AxisTrait.Binary, "Home"),
+
+        // shoulder buttons
+        new(32, AxisTrait.Binary | AxisTrait.LeftSide, "L1"),
+        new(33, AxisTrait.Binary | AxisTrait.RightSide, "R1"),
+        new(34, AxisTrait.Analog | AxisTrait.LeftSide, "L2"),
+        new(35, AxisTrait.Analog | AxisTrait.RightSide, "R2"),
+
+        // gyro
+        new(36, AxisTrait.EulerAngleComponent, "Gyro X"),
+        new(37, AxisTrait.EulerAngleComponent, "Gyro Y"),
+        new(38, AxisTrait.EulerAngleComponent, "Gyro Z"),
+
+        // accelerometer
+        new(39, AxisTrait.Acceleration, "Accelerometer X"),
+        new(40, AxisTrait.Acceleration, "Accelerometer Y"),
+        new(41, AxisTrait.Acceleration, "Accelerometer Z"),
+
+        // battery level
+        new(42, AxisTrait.Analog, "Battery Level"),
+
+        // microphone input level
+        new(43, AxisTrait.Analog, "Microphone Level"),
+    ];
+}
+ ```
+
+As additional food for thought, a Sony Dualshock 2 controller whose face buttons are famously (though infrequently used as) analog, might contain face buttons with the following definition:
+
+```csharp
+_ = new AxisDescription(0, AxisTrait.Binary | AxisTrait.Analog | AxisTrait.RightSide, "Square");
+```
+
+## Axis Groups
+`AxisGroup`s refer to collections of axes whose outputs or layouts are related to one another. If used smartly, they can be used to provide novel high-level remapping schemes, virtual devices, or other features.
+
+### Struct Definition
+
+```csharp
+/// <summary>
+/// This is the declaration of a group of axes within a specific device
+/// </summary>
+/// <param name="Index">Index within the list of <see cref="AxisGroup"/>s</param>
+/// <param name="Purpose">The intended purpose of this group</param>
+/// <param name="Axes">The indices of the axes that make up this group</param>
+/// <param name="TwinIndex">The AxisGroup index of the "twin" of this group - e.g. if this is the left thumbstick, the twin would be the right thumbstick</param>
+/// <param name="Name">A human-readable name for the group</param>
+/// <remarks>Can names be inferred at runtime from the AxisDescription and AxisGroup? Should they be nullable/optional
+/// in the definition and then populated at runtime if they're missing?</remarks>
+public readonly record struct AxisGroup(
+    int Index,
+    AxisGroupType Purpose,
+    IReadOnlyList<int> Axes,

```

Comment:
Should we just do FirstAxisIndex & AxisCount rather than allocating an additional list here? Is there ever a case where a groups axes will not be contiguous?

Reply:
yes I think there is ample opportunity for axes to be non-contiguous. I believe most implementations would be able to be contiguous, but some would be awkward if not impossible. I think WASD would be the most common case where they would likely be non-contiguous, though aside from keyboard inputs it becomes more of a niche issue. But i'd like to avoid closing any doors if i can. 

though I also don't like the need for a collection, i couldn't think of a better way that's still straight-forward

Reply:
Ok